File: kaleido-gradle-plugin/src/main/java/com/tongsr/kaleido/gradle/dsl/KaleidoEscapeHatch.java
package com.tongsr.kaleido.gradle.dsl;

import java.nio.charset.StandardCharsets;
import java.util.Base64;
import org.gradle.api.Named;
import org.gradle.api.provider.Property;
import org.gradle.api.provider.SetProperty;

public abstract class KaleidoEscapeHatch implements Named {
    private final String name;

    protected KaleidoEscapeHatch(String name) {
        this.name = name;
    }

    @Override
    public final String getName() {
        return name;
    }

    public abstract Property<String> getReason();
    public abstract SetProperty<KaleidoProtectionDimension> getDimensions();
    public abstract Property<KaleidoProtectionSelectorKind> getSelectorKind();
    public abstract Property<String> getSelectorValue();

    public final void exact(String identity) {
        getSelectorKind().set(KaleidoProtectionSelectorKind.EXACT);
        getSelectorValue().set(identity);
    }

    public final void prefix(String boundedPrefix) {
        getSelectorKind().set(KaleidoProtectionSelectorKind.PREFIX);
        getSelectorValue().set(boundedPrefix);
    }

    public final String canonicalDeclaration() {
        var dimensions = getDimensions().get().stream().map(Enum::name).sorted().toList();
        return getName() + "|" + getSelectorKind().get().name() + "|"
                + encode(getSelectorValue().get()) + "|" + String.join("+", dimensions)
                + "|" + encode(getReason().get());
    }

    private static String encode(String value) {
        return Base64.getUrlEncoder().withoutPadding()
                .encodeToString(value.getBytes(StandardCharsets.UTF_8));
    }
}

File: kaleido-gradle-plugin/src/main/kotlin/com/tongsr/kaleido/gradle/dsl/KaleidoEscapeHatch.kt
package com.tongsr.kaleido.gradle.dsl

import org.gradle.api.Named
import org.gradle.api.provider.Property
import org.gradle.api.provider.SetProperty
import java.nio.charset.StandardCharsets
import java.util.Base64

abstract class KaleidoEscapeHatch protected constructor(
    private val name: String,
) : Named {
    final override fun getName(): String = name

    abstract val reason: Property<String>
    abstract val dimensions: SetProperty<KaleidoProtectionDimension>
    abstract val selectorKind: Property<KaleidoProtectionSelectorKind>
    abstract val selectorValue: Property<String>

    fun exact(identity: String) {
        selectorKind.set(KaleidoProtectionSelectorKind.EXACT)
        selectorValue.set(identity)
    }

    fun prefix(boundedPrefix: String) {
        selectorKind.set(KaleidoProtectionSelectorKind.PREFIX)
        selectorValue.set(boundedPrefix)
    }

    fun canonicalDeclaration(): String {
        val sortedDimensions = dimensions.get().map(Enum<*>::name).sorted()
        return name + "|" + selectorKind.get().name + "|" +
            encode(selectorValue.get()) + "|" + sortedDimensions.joinToString("+") +
            "|" + encode(reason.get())
    }

    private fun encode(value: String): String = Base64.getUrlEncoder().withoutPadding()
        .encodeToString(value.toByteArray(StandardCharsets.UTF_8))
}
